Add --license option to list-dependencies

ruuda opened the pull request:
Stack makes it a breeze to build a project with lots of (transitive) dependencies without worrying too much about them, but when it comes to distributing the final binary, it is useful to know what your dependencies actually are. In particular, I want to know the licenses of my dependencies, to make sure that they are compatible, and to make sure that I am not accidentally violating one.

For Cabal there is the [`cabal-dependency-licenses`](https://github.com/jaspervdj/cabal-dependency-licenses) package, but I haven’t been able to make it work with Stack. However, because Stack can already list all dependencies, I figured it shouldn’t be too hard to make Stack list the licenses too. So that is what this PR does.

* Add a `--license` flag to `stack list-dependencies`. When specified, `stack list-dependencies` will print the package license instead of the package version.
* As a drive-by change, unify the options for `stack dot` and `stack list-dependencies`. In order not to change the current behaviour,  `--external` defaults to false for `stack dot` and to true for `stack list-dependencies`.

My personal use case for this is that I want to validate the output of `stack list-dependencies --license --no-include-base` against a list of whitelisted licenses as part of CI, to ensure that I don’t accidentally depend on an incompatible library.

Also, I have a question about `--no-include-base`. Would it make sense to have this flag exclude the `template-haskell` and `integer-gmp` packages too? They appear to be “special” just like base.

This is my first time diving into the Stack codebase, so some things can likely be improved. Any feedback would be appreciated.

Blaisorblade commented:
The feature makes sense. The build failures I see come from master (fix should be in #2580).

ruuda commented:
I rebased this on top of the fix.

Blaisorblade commented:
Thanks! Haven't had time to review the code yet though (not sure I'll find it myself, I'm afraid).

borsboom commented on src/Stack/Dot.hs:
```diff
@@ -70,16 +80,37 @@ dot :: (HasEnvConfig env
+  printGraph dotOpts localNames prunedGraph
+
+-- | Information about a package, when available.
+type Payload = (Maybe Version, Maybe License)
```
Any reason not to make this a record type instead?

ruuda replied:
No reason in particular. I’ll change it to a record type as that makes the code below more readable.

borsboom commented:
Overall, LGTM.

ruuda commented:
I rebased this once more on top of master. All commits compile, but they don’t all pass the tests.

@lwm: You moved `dotOptsParser` into `Stack.Options.DockerParser` in #2587, was that by accident? I moved it into a new `Stack.Options.DotParser`.

lwm commented:
@ruuda Yes, that was a mistake. Thank you for catching that and adding the fix!

Blaisorblade commented:
Sorry for not reviewing that more closely (I wanted to merge it before it bitrotted).

ruuda commented:
Rebased once more to resolve changelog conflicts. Are there still any blockers for this?

Blaisorblade commented:
Merged based on the existing review.

ruuda commented:
Thanks!
